Add link to MAX_RETRY allocation explain message (#113657)
elastic/elasticsearch: docs/reference/cluster/allocation-explain.asciidoc

        {
          "decider": "max_retry",
          "decision" : "NO",

>           "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed=true] to retry, and for more information, see https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-allocation-explain.html#_maximum_number_of_retries_exceeded, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"

Reviewer: The API call in the message is actually `POST /_cluster/reroute?retry_failed&metric=none` - see `org.elasticsearch.cluster.routing.allocation.decider.MaxRetryAllocationDecider#RETRY_FAILED_API`.

Author: Thanks, fixed. It’s funny, it looks like the API command as it appears in our documentation right now is actually different from the command that shows up in the in-cluster response message.

>           "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed=true] to retry, and for more information, see https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-allocation-explain.html#_maximum_number_of_retries_exceeded, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"

Reviewer 2: Until we solidify the URL I vote leaving it off. Then Github suggestion to apply Dave's comment I believe would appear as:

```suggestion
          "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed&metric=none] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"
```

        }
      ]
    }
  ]
}
----
// NOTCONSOLE
This message indicates that the cluster was previously unable to
allocate this shard and chose to put a hold on further attempts.
This is done to avoid burdening the cluster with repeated requests that will fail.
If no other `no` decisions are present, then the transient allocation issue
that caused these failures has most likely been resolved, and you can use the

> <<cluster-reroute,the cluster reroute API>> to retry allocation.

Reviewer: I think this will be confusing, there are normally always some `no` decisions e.g. for nodes in the wrong data tier.

Also I'd rather we used the imperative voice: "use the reroute API" rather than just suggesting "you can ...".

Finally there's a duplicate `the` (one inside the link and one outside).

Reviewer 2: Brainstorming, I might say
> Elasticsearch queues shard allocation retries in batches. If there are long running or a high quantity of shard recoveries occurring within the cluster, this process may time out for some shards resulting in `MAX_RETRY`. This surfaces infrequently but is expected to prevent infinite retries which may impact cluster performance. When encountered, run <<cluster-reroute,the cluster reroute API>> to retry allocation.

Author: Changed this to 
> Elasticsearch queues shard allocation retries in batches. If there are long-running shard
recoveries or a high quantity of shard recoveries occurring within the cluster, this
process may time out for some shards, resulting in `max_retry`. This surfaces infrequently
but is expected to prevent infinite retries which may impact cluster performance. When
encountered, run the <<cluster-reroute,cluster reroute>> API to retry allocation.

Which is basically identical but I tweaked the wording on the second sentence because I thought it sounded a bit clearer that way, and also moved 'the' and 'API' out of the link per suggestion from @DaveCTurner 

Thanks!

        {
          "decider": "max_retry",
          "decision" : "NO",

>           "explanation": "shard has exceeded the maximum number of retries [5] on failed allocation attempts - manually call [POST /_cluster/reroute?retry_failed&metric=none] to retry, [unassigned_info[[reason=ALLOCATION_FAILED], at[2024-07-30T21:04:12.166Z], failed_attempts[5], failed_nodes[[mEKjwwzLT1yJVb8UxT6anw]], delayed=false, details[failed shard on node [mEKjwwzLT1yJVb8UxT6anw]: failed recovery, failure RecoveryFailedException], allocation_status[deciders_no]]]"

Reviewer: `org.elasticsearch.cluster.routing.allocation.decider.MaxRetryAllocationDecider#RETRY_FAILED_API` is `POST /_cluster/reroute?retry_failed` in v9, the `metric` parameter is deprecated & ignored. But in v8.x this is correct.

Author: Tweaked this accordingly to just be `POST /_cluster/reroute?retry_failed`

        }
      ]
    }
  ]
}
----
// NOTCONSOLE
Elasticsearch queues shard allocation retries in batches. If there are long-running shard
recoveries or a high quantity of shard recoveries occurring within the cluster, this

> process may time out for some shards, resulting in `max_retry`. This surfaces infrequently

Reviewer: This isn't true, there's no timeout in play here. You need to get 5 genuine failures in a row before you see this.
